File: TracePluginDetail/src/main/java/com/zz/trace/ZTracePlugin.java
package com.zz.trace;

import com.android.build.gradle.AppExtension;

import org.gradle.api.Plugin;
import org.gradle.api.Project;

/**
 * @author zz
 * @date 2023/4/11
 * @describe
 **/
public class ZTracePlugin implements Plugin<Project> {
    @Override
    public void apply(Project project) {
        System.out.println("===========  trace plugin start  ============");
        AppExtension appExtension = project.getExtensions().findByType(AppExtension.class);
        assert appExtension != null;
        appExtension.registerTransform(new ZTraceTransform());
    }
}

File: TracePluginDetail/src/main/kotlin/com/zz/trace/ZTracePlugin.kt
package com.zz.trace

import com.android.build.gradle.AppExtension
import org.gradle.api.GradleException
import org.gradle.api.Plugin
import org.gradle.api.Project

class ZTracePlugin : Plugin<Project> {
    companion object {
        const val TAG = "Trace plugin"
    }

    override fun apply(p0: Project) {
        println("===========  开始  ============")
        p0.plugins.forEach {
            println(it.toString())
        }
//        p0.pluginManager.withPlugin("com.android.build.gradle.internal.plugins.AppPlugin") {
//            val extension = p0.extensions.getByType(AppExtension::class.java)
//            extension.applicationVariants.forEach {
//                println("name: ${it.name}")
//            }
//        }
//        if (p0.plugins.hasPlugin("com.android.application")) {
//            throw GradleException("Matrix Plugin, Android Application plugin required.")
//        }
        p0.extensions.getByType(AppExtension::class.java).registerTransform(TraceTransform())
        println("===========  结束  ============")
    }
}